### PR TITLE
feat: add cluster_provider_resource_quota_consumed_ metric

### PR DESCRIPTION
**Pull request by JasonPowr**
## Description
Adds kas_fleet_manager_quota_consumed to the KAS Fleet Manager metrics. This represents the quota currently used by the organisation of the ocm account used by KAS Fleet Manager.

Jira: [Here](https://issues.redhat.com/browse/MGDSTRM-9404)

## Verification Steps
1.Run kas fleet manager with ocm credentials specified
2.Curl http://localhost:8080/metrics
3.Ensure the quota metric as defined above are included in the response with correct values

## Checklist (Definition of Done)

- [x] All acceptance criteria specified in JIRA have been completed
- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~~Documentation added for the feature~~
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [x] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] ~~Required metrics/dashboards/alerts have been added (or PR created).~~
- [ ] ~~Required Standard Operating Procedure (SOP) is added.~~
- [ ] ~~JIRA has been created for changes required on the client side~~

### CODE REVIEW

**Comment by JasonPowr** [comment]
@JameelB  :)

File: internal/kafka/internal/workers/clusters_mgr.go [hunk 1]
```diff
@@ -1245,3 +1248,22 @@ func (c *ClusterManager) setKafkaPerClusterCountMetrics() error {
 
 	return nil
 }
+
+func (c *ClusterManager) setKasFleetManagerClusterProviderResourceQuotaConsumedMetric() error {
+	ocmCredentialsSpecified := (c.OCMConfig.ClientID != "" && c.OCMConfig.ClientSecret != "") || c.OCMConfig.SelfToken != ""
```
Comment (machi1990) [inline internal/kafka/internal/workers/clusters_mgr.go:1253]: This is also here https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1251 maybe it is good to have a `.CheckIfCredentialsAreProvided()` method in the `OCMConfig` struct that can be used here and in the linked PR

/cc @JameelB @JasonPowr

File: internal/kafka/test/integration/cloudprovider_test.go [hunk 2]
```diff
@@ -360,9 +363,18 @@ func TestListCloudProviderRegions(t *testing.T) {
...
 	})
 	defer teardown()
+	var clusterService services.ClusterService
+	h.Env.MustResolve(&clusterService)
+	cluster := dataPlaneClusterConfig.ClusterConfig.GetManualClusters()
+
+	for _, c := range cluster {
+		_, err := common.WaitForClusterStatus(h.DBFactory(), &clusterService, c.ClusterId, api.ClusterReady)
+		g.Expect(err).NotTo(gomega.HaveOccurred())
+	}
```
Comment (machi1990) [inline internal/kafka/test/integration/cloudprovider_test.go:377]: Why is this test modified?

File: internal/kafka/test/integration/kafkas_test.go [hunk 3]
```diff
@@ -774,6 +774,11 @@ func TestKafkaCreate_TooManyKafkas(t *testing.T) {
 	h, client, teardown := test.NewKafkaHelperWithHooks(t, ocmServer, configHook)
 	defer teardown()
 
+	var clusterService services.ClusterService
+	h.Env.MustResolve(&clusterService)
+	_, err := common.WaitForClusterStatus(h.DBFactory(), &clusterService, "test01", api.ClusterReady)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
```
Comment (machi1990) [inline internal/kafka/test/integration/kafkas_test.go:780]: Why is this modified?

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -256,6 +275,16 @@ var clusterStatusCapacityMaxMetric = prometheus.NewGaugeVec(
 	clusterStatusCapacityLabels,
 )
 
+// create a new gaugeVec for the quota an organisation is currently using
+var KasFleetManagerClusterProviderResourceQuotaConsumedMetric = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Subsystem: KasFleetManager,
+		Name:      KasFleetManagerClusterProviderResourceQuotaConsumed,
+		Help:      "quota an organisation is currently using",
```
Comment (machi1990) [inline pkg/metrics/metrics.go:283]:
```suggestion
		Help:      "cluster resource quota that are current consumed",
```

File: internal/kafka/internal/workers/clusters_mgr.go (hunk 1 above)
Comment (JameelB) [reply internal/kafka/internal/workers/clusters_mgr.go:1253]: having a method for it sounds good. We can use it in other places too if needed in the future.

**Comment by JameelB** [comment]
Majority of the changes here will be the same for https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1251. Looking to get this in first then I can rebase my changes once merged.

File: internal/kafka/internal/workers/clusters_mgr.go [hunk 5]
```diff
@@ -1245,3 +1248,22 @@ func (c *ClusterManager) setKafkaPerClusterCountMetrics() error {
 
 	return nil
 }
+
+func (c *ClusterManager) setKasFleetManagerClusterProviderResourceQuotaConsumedMetric() error {
```
Comment (JameelB) [inline internal/kafka/internal/workers/clusters_mgr.go:1252]: Can we change the name to this? 
```suggestion
func (c *ClusterManager) setClusterProviderResourceQuotaMetrics() error {
```

I don't think we need separate methods betwen max allowed and consumed since both requires the same logic. The max allowed metric updated can be added in here as well.

File: internal/kafka/internal/workers/clusters_mgr.go (hunk 1 above)
Comment (miguelsorianod) [reply internal/kafka/internal/workers/clusters_mgr.go:1253]: Shouldn't we check instead whether the cluster provider is OCM first at all?

Comment (miguelsorianod) [reply internal/kafka/internal/workers/clusters_mgr.go:1253]: Additionally, why is this metric limited to OCM?

File: internal/kafka/test/integration/cloudprovider_test.go (hunk 2 above)
Comment (JasonPowr) [reply internal/kafka/test/integration/cloudprovider_test.go:377]: Hey @machi1990  After making the changes for this PR These two integration tests started to fail. I believe there was some sort of timing issue that was causing them to fail and @JameelB  helped me fix them.

File: internal/kafka/internal/workers/clusters_mgr.go (hunk 1 above)
Comment (machi1990) [reply internal/kafka/internal/workers/clusters_mgr.go]: > Shouldn't we check instead whether the cluster provider is OCM instead?

These are not cluster specific metrics but cluster provider e.g OCM, wide metrics.  

> Additionally, why is this metric limited to clusters provisioned through OCM?

For other only provider that we've i.e `standalone` the count will be zero everywhere (`consumed` and `max`). I think once we've other cluster providers we could add them.

File: internal/kafka/test/integration/cloudprovider_test.go (hunk 2 above)
Comment (machi1990) [reply internal/kafka/test/integration/cloudprovider_test.go:377]: Thanks

File: internal/kafka/internal/workers/clusters_mgr.go [hunk 6]
```diff
@@ -1245,3 +1248,22 @@ func (c *ClusterManager) setKafkaPerClusterCountMetrics() error {
 
 	return nil
 }
+
+func (c *ClusterManager) setClusterProviderResourceQuotaMetrics() error {
+	ocmCredentialsSpecified := c.OCMConfig.CheckIfCredentialsAreProvided()
+	if ocmCredentialsSpecified {
```
Comment (machi1990) [inline internal/kafka/internal/workers/clusters_mgr.go:1254]: we could do an early return here by inversing the condition.
```suggestion
	if !ocmCredentialsSpecified {
```

what do you think? @JameelB @JasonPowr

File: pkg/client/ocm/config.go [hunk 7]
```diff
@@ -63,6 +63,13 @@ func (c *OCMConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.ClusterLoggingOperatorAddonID, "cluster-logging-operator-addon-id", "", "The name of the cluster logging operator addon. An empty string indicates that the operator should not be installed")
 }
 
+func (c *OCMConfig) CheckIfCredentialsAreProvided() bool {
+	if (c.ClientID != "" && c.ClientSecret != "") || c.SelfToken != "" {
+		return true
+	}
+	return false
```
Comment (machi1990) [inline pkg/client/ocm/config.go:70]:
```suggestion
	return c.ClientID != "" && c.ClientSecret != "") || c.SelfToken != ""
```

File: pkg/client/ocm/config.go [hunk 8]
```diff
@@ -63,6 +63,13 @@ func (c *OCMConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.ClusterLoggingOperatorAddonID, "cluster-logging-operator-addon-id", "", "The name of the cluster logging operator addon. An empty string indicates that the operator should not be installed")
 }
 
+func (c *OCMConfig) CheckIfCredentialsAreProvided() bool {
```
Comment (machi1990) [inline pkg/client/ocm/config.go:66]: we can add a unit test for this?

**Review by machi1990 — APPROVED** [review]
the code change lgtm

I've left some minor suggestions, once these are addressed.  
I'll leave it out to @JameelB @miguelsorianod to see if their reviews have been addressed and merge

File: internal/kafka/test/integration/cloudprovider_test.go (hunk 2 above)
Comment (JameelB) [reply internal/kafka/test/integration/cloudprovider_test.go:377]: It's because this change adds additional time to the processMetrics which is the first step in the cluster manager reconcile. This results in the dummy clusters to not be recorded in the database in time for the assertions in this test. 

Having to wait for the cluster to be ready and recorded in the database makes sense here either way since the cloud provider regions endpoint relies on a cluster being there, otherwise it the cloud provider regions endpoint returns an empty list.

File: internal/kafka/internal/workers/clusters_mgr.go (hunk 6 above)
Comment (JameelB) [reply internal/kafka/internal/workers/clusters_mgr.go]: Yeah that makes sense. 

Just note that this will need to be modified if we ever add new providers that we want to expose this metric for since we wouldn't want to return straight away in that case. This is probably far in the future though :)

File: internal/kafka/test/integration/cloudprovider_test.go [hunk 9]
```diff
@@ -360,9 +363,18 @@ func TestListCloudProviderRegions(t *testing.T) {
...
 		})
+		dataPlaneClusterConfig = dc
 
 	})
 	defer teardown()
+	var clusterService services.ClusterService
+	h.Env.MustResolve(&clusterService)
+	cluster := dataPlaneClusterConfig.ClusterConfig.GetManualClusters()
+
+	for _, c := range cluster {
```
Comment (JameelB) [inline internal/kafka/test/integration/cloudprovider_test.go:374]: let's make it clear that this is a slice
```suggestion
	clusters := dataPlaneClusterConfig.ClusterConfig.GetManualClusters()

	for _, c := range clusters {
```

File: pkg/metrics/metrics.go [hunk 10]
```diff
@@ -73,6 +73,11 @@ const (
 	// ClusterStatusCapacityAvailable - metric name for the number of available instances
 	ClusterStatusCapacityAvailable = "cluster_status_capacity_available"
 
+	// KasFleetManagerClusterProviderResourceQuotaConsumed - metric name for how much quota an organisation is currently using
+	KasFleetManagerClusterProviderResourceQuotaConsumed = "kas_fleet_manager_cluster_provider_resource_quota_consumed"
```
Comment (JameelB) [inline pkg/metrics/metrics.go:77]: No need to add KasFleetManager as a prefix. This is taken care of by `Subsystem: KasFleetManager,` in the metric definition. 

```suggestion
	// KasFleetManagerClusterProviderResourceQuotaConsumed - metric name for how much quota an organisation is currently using
	ClusterProviderResourceQuotaConsumed = "cluster_provider_resource_quota_consumed"
```

File: pkg/metrics/metrics.go [hunk 11]
```diff
@@ -73,6 +73,11 @@ const (
...
+
+	LabelQuotaId    = "quota_id"
+	LabelProvider   = "provider"
```
Comment (JameelB) [inline pkg/metrics/metrics.go:80]: Let's make this more specific to cluster_provider to differentiate it with other provider types like cloud provider.

File: pkg/metrics/metrics.go [hunk 12]
```diff
@@ -158,6 +163,11 @@ var clusterStatusCapacityLabels = []string{
 	LabelCloudProvider,
 }
 
+var KasFleetManagerClusterProviderResourceQuotaConsumedLabels = []string{
```
Comment (JameelB) [inline pkg/metrics/metrics.go:166]: Remove KasFleetManager prefix to be consistent with other variables. 
```suggestion
var ClusterProviderResourceQuotaConsumedLabels = []string{
```

File: pkg/metrics/metrics.go [hunk 13]
```diff
@@ -73,6 +73,11 @@ const (
 	// ClusterStatusCapacityAvailable - metric name for the number of available instances
 	ClusterStatusCapacityAvailable = "cluster_status_capacity_available"
 
+	// KasFleetManagerClusterProviderResourceQuotaConsumed - metric name for how much quota an organisation is currently using
```
Comment (JameelB) [inline pkg/metrics/metrics.go:76]:
```suggestion
	// KasFleetManagerClusterProviderResourceQuotaConsumed - metric name for how much quota, given to a user by a cluster provider, is currently used.
```

quota by organisation is an ocm detail and other providers may not implement it this way.

File: pkg/metrics/metrics.go [hunk 14]
```diff
@@ -214,6 +224,15 @@ func UpdateClusterStatusCapacityMaxCount(provider string, region, instanceType,
 	clusterStatusCapacityMaxMetric.With(labels).Set(count)
 }
 
+// UpdateKasFleetManagerClusterProviderResourceQuotaConsumed - sets quota an organisation is currently using
+func UpdateKasFleetManagerClusterProviderResourceQuotaConsumed(quotaId, provider string, count int) {
```
Comment (JameelB) [inline pkg/metrics/metrics.go:228]: Move this closer to the metric definition it's updating. Best to keep them together to find related uses of the metric easier.

Comment (JameelB) [reply pkg/metrics/metrics.go:228]: update the description to replace `organisation` by `user`.

[optional] change `provider` -> `clusterProvider` to make it clear what provider this method is looking for.

File: pkg/client/ocm/config.go [hunk 15]
```diff
@@ -63,6 +63,10 @@ func (c *OCMConfig) AddFlags(fs *pflag.FlagSet) {
...
```
Comment (ziccardi) [inline pkg/client/ocm/config.go:66]: Function that just returns a boolean, are usually used to answer to a specific question with a yes or no. IMO this would be better called something like `HasCredentials`, 'CredentialsExists`, `AreCredentialsProvided` or something like that.

**Review by JameelB — APPROVED** [review]
